\documentclass[10pt,letterpaper]{article}
\usepackage[top=0.85in,left=2.75in,footskip=0.75in,marginparwidth=2in]{geometry}

% use Unicode characters - try changing the option if you run into troubles with special characters (e.g. umlauts)
\usepackage[utf8]{inputenc}

% clean citations
\usepackage{cite}

% hyperref makes references clicky. use \url{www.example.com} or \href{www.example.com}{description} to add a clicky url
\usepackage{nameref,hyperref}

% line numbers
\usepackage[right]{lineno}

% improves typesetting in LaTeX
\usepackage{microtype}
\DisableLigatures[f]{encoding = *, family = * }

% text layout - change as needed
\raggedright
\setlength{\parindent}{0.5cm}
\textwidth 5.25in 
\textheight 8.75in

% Remove % for double line spacing
%\usepackage{setspace} 
%\doublespacing

% use adjustwidth environment to exceed text width (see examples in text)
\usepackage{changepage}

% adjust caption style
\usepackage[aboveskip=1pt,labelfont=bf,labelsep=period,singlelinecheck=off]{caption}

% remove brackets from references
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}
\makeatother

% headrule, footrule and page numbers
\usepackage{lastpage,fancyhdr,graphicx}
\usepackage{epstopdf}
\pagestyle{myheadings}
\pagestyle{fancy}
\fancyhf{}
\rfoot{\thepage/\pageref{LastPage}}
\renewcommand{\footrule}{\hrule height 2pt \vspace{2mm}}
\fancyheadoffset[L]{2.25in}
\fancyfootoffset[L]{2.25in}

% use \textcolor{color}{text} for colored text (e.g. highlight to-do areas)
\usepackage{color}

% define custom colors (this one is for figure captions)
\definecolor{Gray}{gray}{.25}

% this is required to include graphics
\usepackage{graphicx}

% use if you want to put caption to the side of the figure - see example in text
\usepackage{sidecap}

% use for have text wrap around figures
\usepackage{wrapfig}
\usepackage[pscoord]{eso-pic}
\usepackage[fulladjust]{marginnote}
\reversemarginpar

% Adding multirow.
\usepackage{multirow}

% Other required things:
\usepackage{color}
\usepackage{subcaption}
\captionsetup[subfigure]{justification=centering}
\usepackage{amsmath}
\newcommand\code[1]{{\small\texttt{#1}}}

% document begins here
\begin{document}
\vspace*{0.35in}

% title goes here:
\begin{flushleft}
{\Large
    \textbf\newline{Using an appropriate pseudo-count for log-transformation of normalized single-cell RNA sequencing data}
}
\newline

% authors go here:
%\\
Aaron Lun\textsuperscript{1,*}
\\
\bigskip
\bf{1} Cancer Research UK Cambridge Institute, University of Cambridge, Li Ka Shing Centre, Robinson Way, Cambridge CB2 0RE, United Kingdom \\
\bigskip

\end{flushleft}

\section{Background}
Log-transformed expression values are widely used in analyses of single-cell RNA sequencing (scRNA-seq) and other transcriptomic data.
This is driven the simplicity of the log-transformation and its moderate variance stabilizing capabilities on many types of non-negative data.
In particular, the log-transformation reduces the impact of stochastic fluctuations in the counts for high-abundance genes.
This would otherwise result in large differences in the counts that are mostly uninteresting as the fold changes are small.
Differences between log-values are also approximately interpretable as log-fold changes between cells, which are often more relevant than differences in the counts.
This is useful for ensuring that the relative rather than absolute differences in the counts are used in distance-based procedures like clustering or trajectory construction.

Despite its popularity, the log-transformation has a number of problems such as incomplete variance stabilization and arbitrariness in the choice of pseudo-count.
One issue of particular interest is that the mean of the log-counts is not generally the same as the log-mean count \cite{hicks2017missing}.
This is problematic in (sc)RNA-seq contexts where the log-transformation is applied to normalized expression data.
Here, normalization is performed to remove inter-sample biases on the count scale \cite{robinson2010scaling,lun2016pooling}.
For genes that are not differentially expressed (DE), the expectation of the normalized expression is the same between samples.
However, the expectation of the log-normalized values may not be the same, resulting in spurious differences between samples/cells in the log-scale.

In this report, we describe the nature and impact of the discrepancy between log-mean and mean-log values.
We also show how the added pseudo-count can be increased to cap the discrepancy at the cost of reducing interpretability.

\section{Quantifying the discrepancy}
Let $X_i$ denote a random variable for a non-negative count of a gene $g$ in cell $i$.
The log-transformed normalized expression value is defined as 
\[
Z_i = \log(X_i / s_i + c) \;,
\]
where $s_i$ is the size factor for $i$ and $c$ is the pseudo-count.
(We will omit the gene indexing for clarity, as we are only referring to a single gene throughout.)
The second-order Taylor series approximation for the expectation of $Z_i$ is
\[
E(Z_i) \approx \log[E(X_i/s_i) + c] - \frac{\mbox{var}(X_i)s_i^{-2}}{2[E(X_i/s_i) + c]^2} \;.
\]
The second term represents the discrepancy between the mean-log and the log-mean.

Now, consider two cells $i=1$ and $i=2$ that differ only in their $s_i$.
Assume that $g$ is non-DE between these two cells, such that $E(X_i/s_i)=\mu$ for both.
This means that
\[
E(Z_i) \approx \log(\mu + c) - \frac{\mbox{var}(X_i)s_i^{-2}}{2(\mu + c)^2} \;.
\]
The difference in $E(Z_i)$ between these two cells is interpreted as the expected log-fold change in the normalized counts, and can be written as
\begin{equation}
\Delta_{12} \approx \frac{\mbox{var}(X_1)s_1^{-2} -  \mbox{var}(X_2)s_2^{-2}}{2(\mu + c)^2} \;. \label{eqn:spuriousdiff}
\end{equation}
Any non-zero $\Delta_{12}$ is a spurious difference as the log-fold change between cells should be zero after normalization.
In general, $\Delta_{12}\neq 0$ due to differences in the size factors and subsequently variances at different points in the mean-variance relationship.

We performed simulations to quantify this spurious difference in a range of scenarios for the typical $c=1$ (Figure~\ref{fig:maxeffect}).
The difference was most prominent when the size factors were different and either or both of them were less than unity.
This is consistent with Equation~\ref{eqn:spuriousdiff}, where $\Delta_{12}$ increases in magnitude with smaller $s_i$.
We note that 10-fold difference in the size factors across cells is not unusual in real scRNA-seq data \cite{lun2016pooling} due to the variability of capture efficiency and amplification.
However, even in this situation, we can obtain a log-difference of 1.
This corresponds to an artificial 2-fold change in expression, which can result in misleading biological conclusions.

\begin{figure}[btp]
\centering
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 5mm,clip,page=1]{../scripts/pics/max_effect.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 5mm,clip,page=2]{../scripts/pics/max_effect.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 5mm,clip,page=3]{../scripts/pics/max_effect.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 5mm,clip,page=4]{../scripts/pics/max_effect.pdf}
    \caption{}
\end{subfigure}
\caption{Maximum difference in the log-expression values for non-DE simulated data.
Each difference was computed between the average $Z_i$ of 10000 counts for $i=1$ and 2, using $c=1$.
Counts were sampled from a negative binomial distribution with mean $\mu s_i$ and the specified dispersion $\varphi$.
For each combination of the first size factor ($s_1$), second size factor ($s_2$) and $\varphi$, the maximum difference was determined across all $\mu \in [10^{-3}, 10^3]$.
Maximum differences are only shown for scenarios where $s_1 \ge s_2$, for simplicity.
}
\label{fig:maxeffect}
\end{figure}

\section{Exploring alternative transformations}
Spurious differences for non-DE genes may also be present with other transformations.
We considered the square root, which provides good variance stabilization for Poisson-distributed counts;
and the variance stabilizing transformation (VST) for negative binomial (NB)-distributed counts from \emph{DESeq2} \cite{love2014moderated}.
In simulations of non-DE genes, we observed large differences in the mean of transformed values with both transformations (Figure~\ref{fig:alttransform}).
This indicates that variance stabilization does not protect against spurious differences in the mean.
Perhaps no single transformation exists that 
(i) yields transformed values that are easily interpretable,
(ii) provides complete variance stabilization, and
(iii) avoids spurious differences in the mean of transformed values.

\begin{figure}
\centering
\begin{subfigure}[b]{0.32\textwidth}
    \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 5mm,clip,page=2]{../scripts/pics/alternative_pois.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.32\textwidth}
    \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 5mm,clip,page=1]{../scripts/pics/alternative_pois.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.32\textwidth}
    \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 5mm,clip,page=3]{../scripts/pics/alternative_pois.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.32\textwidth}
    \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 5mm,clip,page=2]{../scripts/pics/alternative_nb.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.32\textwidth}
    \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 5mm,clip,page=1]{../scripts/pics/alternative_nb.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.32\textwidth}
    \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 5mm,clip,page=3]{../scripts/pics/alternative_nb.pdf}
    \caption{}
\end{subfigure}
\caption{Distribution of mean expression values for Poisson-distributed counts (a, b, c) or NB-distributed counts (d, e, f) after applying a variety of transformations.
Simulated count data was generated for 10000 genes in 500 cells, with the mean count set to 1 for half of the cells (smaller group) and 100 for the other half (larger group).
The dispersion was set to 1 for the NB simulations.
Size factors were calculated by scaling the library sizes to a mean of unity across cells.
Transformations were applied to the size factor-scaled counts, using a pseudo-count of 1 for the log$_2$-transformation and \code{fitType="mean"} for the \textit{DESeq2} VST.
Boxplots represent the distribution of mean transformed values across genes, with outlier genes shown as separate dots.
}
\label{fig:alttransform}
\end{figure}

\section{Computing a suitable pseudo-count}
We assume that $X_i$ follows a negative binomial (NB) distribution with mean $s_i\mu$ and dispersion $\varphi$.
This means that we can simplify the expression for $\Delta_{12}$ to
\begin{align*}
\Delta_{12} 
&\approx \frac{\mu s_1^{-1} + \varphi \mu^2}{2(\mu + c)^2} - \frac{\mu s_2^{-1} + \varphi \mu^2}{2(\mu + c)^2} \\ 
&= \frac{\mu (s_1^{-1} - s_2^{-1})}{2(\mu + c)^2} \;.
\end{align*}
This spurious difference in the log-expression values is maximized when $\mu = c$, yielding 
\[
\tilde\Delta_{12} = \frac{(s_1^{-1} - s_2^{-1})}{8c} \;.
\]
Thus, we can cap the maximum error $\tilde\Delta_{12}$ regardless of the values of $s_i$ by defining
\[
    c \propto |s_1^{-1} - s_2^{-1}| \;.
\]
If we set the proportionality constant to unity, the maximum log-difference should be 0.125 (or $[8\log(2)]^{-1} \approx 0.18$, on the $\log_2$-scale).
We observe this in a range of simulation scenarios (Figure~\ref{fig:cappederr}), indicating that our approximate upper bound is correct. 

\begin{figure}[btp]
\centering
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 5mm,clip,page=1]{../scripts/pics/error_bound.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 5mm,clip,page=2]{../scripts/pics/error_bound.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 5mm,clip,page=3]{../scripts/pics/error_bound.pdf}
    \caption{}
\end{subfigure}
\begin{subfigure}[b]{0.49\textwidth}
    \includegraphics[width=\textwidth,trim=0mm 5mm 0mm 5mm,clip,page=4]{../scripts/pics/error_bound.pdf}
    \caption{}
\end{subfigure}
\caption{Maximum difference in the log-expression values for non-DE simulated data after setting $c = |s_1^{-1} - s_2^{-1}|$.
Each difference was computed between the average $Z_i$ of 10000 counts for $i=1$ and 2.
Counts were sampled from a negative binomial distribution with mean $\mu s_i$ and the specified dispersion $\varphi$.
For each combination of the first size factor ($s_1$), second size factor ($s_2$) and $\varphi$, the maximum difference was determined across all $\mu \in [10^{-3}, 10^3]$.
Results are not shown for scenarios where $s_1=s_2$ as no difference in log-expression is expected (dashed lines connect adjacent points for ease of visualization).
}
\label{fig:cappederr}
\end{figure}

In practice, downstream procedures such as clustering will compare many cells to one another instead of just two specific cells.
The exact choices of $s_1$ and $s_2$ are not obvious when considering the calculation of a single pseudo-count for the entire data set.
A simple strategy is to set $s_1$ to the smallest size factor and $s_2$ to the largest size factor
(or robust equivalents thereof, e.g., the 5\textsuperscript{th} and 95\textsuperscript{th} percentiles).
This ensures that the difference in log-values for a non-DE gene does not exceed the theoretical upper bound for any pair of cells.
We thus define our empirical pseudo-count as
\[
\hat c = \max\{1, |s_{min}^{-1} - s_{max}^{-1}|\} \;,
\]
which ensures that the pseudo-count is at least unity when all size factors are equal.

\section{Consequences of increasing the pseudo-count}
A larger $c$ will shrink the differences in the log-transformed values towards zero for all genes.
This mitigates the effect of spurious differences in non-DE genes but will also compromise interpretation of log-fold changes for genuine DE genes.
A near-zero difference between subpopulations for low-abundance genes cannot be treated as an accurate estimate of the true log-fold change. 
This is obviously problematic if important changes in expression of low-abundance genes are overlooked due to the small magnitude of the difference in log-values.
Greater shrinkage also means that distance calculations are driven by moderate- to high-abundance genes.
This may affect resolution of biological signal that is only present in low-abundance genes.
That said, detection power for DE genes in a formal hypothesis testing framework is mostly unaffected (Figure~\ref{fig:power}).
This is because the larger $c$ also reduces the variance in the log-transformed values, 
thus compensating for the decrease in the mean difference.

\begin{figure}
\begin{center}
\includegraphics[width=\textwidth]{../scripts/pics/power.pdf}
\end{center}
\caption{Welch $t$-statistics as a function of the DE fold change, for log-transformed data with a variety of added pseudo-counts.
We considered an experimental design with 1000 cells in each group.
Counts were sampled from a NB distribution with a mean of $\mu$ in one group and $f\mu$ in the other group for some fold change $f$.
Log$_2$-transformed values were computed with a range of pseudo-counts, and a Welch $t$-statistic was computed for the difference in transformed values between groups.
All size factors were set to unity during the transformation, to reflect the fact that $f$ represents geniune DE. 
}
\label{fig:power}
\end{figure}


\bibliography{ref}
\bibliographystyle{unsrt}

\end{document}
